src/defrag.c

```diff
@@ -840,299 +914,450 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
-    } while (1);
-}
+        }
 
-#define INTERPOLATE(x, x1, x2, y1, y2) ((y1) + ((x) - (x1)) * ((y2) - (y1)) / ((x2) - (x1)))
-#define LIMIT(y, min, max) ((y) < (min) ? min : ((y) > (max) ? max : (y)))
+        if (timeout) break;
 
-/* decide if defrag is needed, and at what CPU effort to invest in it */
-void computeDefragCycles(void) {
-    size_t frag_bytes;
-    float frag_pct = getAllocatorFragmentation(&frag_bytes);
-    /* If we're not already running, and below the threshold, exit. */
-    if (!server.active_defrag_running) {
-        if (frag_pct < server.active_defrag_threshold_lower || frag_bytes < server.active_defrag_ignore_bytes) return;
+        if (defrag_later_cursor == 0) {
+            // the item is finished, move on
+            listDelNode(defrag_later, head);
+        }
+
+        if (++iterations > 16 || server.stat_active_defrag_hits - prev_defragged > 512 ||
+            server.stat_active_defrag_scanned - prev_scanned > 64) {
+            if (getMonotonicUs() > endtime) break;
+            iterations = 0;
+            prev_defragged = server.stat_active_defrag_hits;
+            prev_scanned = server.stat_active_defrag_scanned;
+        }
     }
 
-    /* Calculate the adaptive aggressiveness of the defrag based on the current
-     * fragmentation and configurations. */
-    int cpu_pct = INTERPOLATE(frag_pct, server.active_defrag_threshold_lower, server.active_defrag_threshold_upper,
-                              server.active_defrag_cycle_min, server.active_defrag_cycle_max);
-    cpu_pct = LIMIT(cpu_pct, server.active_defrag_cycle_min, server.active_defrag_cycle_max);
+    return (!defrag_later || listLength(defrag_later) == 0) ? DEFRAG_DONE : DEFRAG_NOT_DONE;
+}
 
-    /* Normally we allow increasing the aggressiveness during a scan, but don't
-     * reduce it, since we should not lower the aggressiveness when fragmentation
-     * drops. But when a configuration is made, we should reconsider it. */
-    if (cpu_pct > server.active_defrag_running || server.active_defrag_configuration_changed) {
-        server.active_defrag_running = cpu_pct;
-        server.active_defrag_configuration_changed = 0;
-        serverLog(LL_VERBOSE, "Starting active defrag, frag=%.0f%%, frag_bytes=%zu, cpu=%d%%", frag_pct, frag_bytes,
-                  cpu_pct);
+
+/* This helper function handles most of the work for iterating over a kvstore.  'privdata', if
+ * provided, MUST begin with 'kvstoreIterState' and this part is automatically updated by this
+ * function during the iteration. */
+static doneStatus defragStageKvstoreHelper(monotime endtime,
+                                           kvstore *kvs,
+                                           dictScanFunction scan_fn,
+                                           kvstoreHelperPreContinueFn precontinue_fn,
+                                           const dictDefragFunctions *defragfns,
+                                           void *privdata) {
```

Comment:
I see that `privdata` internal value is set in this function but it's not really used cross iterations or by caller in the upper level, so why are you setting the state in privdata? why not just pass the state to `precontinue_fn`? and not set it on privdata?

Reply:
1. the caller might use this in the future
2. the caller needs to send privdata to the precontinue function also

So, we could pass both the `kvstoreIterState` and the `privdata` to the precontinue function separately, but that wouldn't allow potential future access by the caller.